beginner_source/onnx/onnx_registry_tutorial.py

```diff
@@ -0,0 +1,469 @@
+# -*- coding: utf-8 -*-
+
+"""
+`Introduction to ONNX <intro_onnx.html>`_ ||
+`Export a PyTorch model to ONNX <export_simple_model_to_onnx_tutorial.html>`_ ||
+**Introduction to ONNX Registry**
+
+Introduction to ONNX Registry
+=============================
+
+**Authors:** Ti-Tai Wang (titaiwang@microsoft.com)
+"""
+
+
+###############################################################################
+# Overview
+# ~~~~~~~~
+#
+# This tutorial is an introduction to ONNX registry, which
+# empowers users to create their own ONNX registries enabling
+# them to address unsupported operators in ONNX.
+#
+# In this tutorial, we will cover the following scenarios:
+#
+# * Unsupported ATen operators
+# * Unsupported ATen operators with existing ONNX Runtime support
+# * Unsupported PyTorch operators with no ONNX Runtime support
+#
+
+
+import torch
+print(torch.__version__)
+torch.manual_seed(191009)  # set the seed for reproducibility
+
+import onnxscript  # pip install onnxscript
+print(onnxscript.__version__)
+
+# NOTE: opset18 is the only version of ONNX operators we are
+# using in torch.onnx.dynamo_export for now.
+from onnxscript import opset18
+
+import onnxruntime  # pip install onnxruntime
+print(onnxruntime.__version__)
+
+
+######################################################################
+# Unsupported ATen operators
+# ---------------------------------
+#
+# ATen operators are implemented by PyTorch, and the ONNX exporter team must manually implement the
+# conversion from ATen operators to ONNX operators through [ONNX Script](https://onnxscript.ai/). Although the ONNX exporter
+# team has been making their best efforts to support as many ATen operators as possible, some ATen
+# operators are still not supported. In this section, we will demonstrate how you can implement any
+# unsupported ATen operators, which can contribute back to the project through the PyTorch GitHub repo.
+#
+# If the model cannot be exported to ONNX, for instance, :class:`aten::add.Tensor` is not supported
+# by ONNX The error message can be found, and is as follows (for example,  ``aten::add.Tensor``):
+#    ``RuntimeErrorWithDiagnostic: Unsupported FX nodes: {'call_function': ['aten.add.Tensor']}. ``
```

Comment:
```suggestion
# In this section, we will assume that `aten::add.Tensor` is not supported by the ONNX registry, and we will demonstrate how to support it.
# When a model cannot be exported to ONNX due to a missing operator, the ONNX exporter will show an error message with something similar to:
#
# .. code-block:: python
#    RuntimeErrorWithDiagnostic: Unsupported FX nodes: {'call_function': ['aten.add.Tensor']}.
```

Reply:
This section is focussed on the adding missing aten operators, but the example actually replaces an existing one, which is the topic of the next section.

I think we should add support to a truly missing operator? Maybe a variant of an existing operator that we don't really care about in having in the official registry?

I am trying to prevent confusion between the 3 scenarios. The missing operator section is using a trick of replacing existing operator, which beginners might find confusing

Reply:
I do agree 100% with you that we should not miss an aten operator on purpose just for the sake of the tutorial. I would never suggest that :)

In fact, I am Ok in keeping the current approach if we cannot find another way that frees user from hacking the registry themselves. In that case, I would create a new section/snippet block just for the un-registration part and make it part of the tutorial. That would 1) teach how to remove an operator from registry and 2) create a "physical" separation between the "adding" new operator from the "removing" the operator in the tutorial.

Nonetheless, I would like to inquire about options that make things simpler for the user.

For example, can we leverage the matching logic to register an equivalent operator with different names(overloads)? Maybe leveraging the `default` overload or another overload the . Can we try to register a `aten::add.Tensor.default` that we don't implement today because of our onnx-fx interpreter can match `aten::add.Tensor.default` to `aten::add.Tensor` that already exists?

Maybe something similar to what is being discussed at https://github.com/pytorch/pytorch/pull/109966 in which `aten::split(tensor,dim)` already exists but a `aten::split(tensor, dim, drop_remainder=False)` is added with the same behavior? #109966 could be improved to make this kind of scenario possible, if not already

Reply:
> For example, can we leverage the matching logic to register an equivalent operator with different names(overloads)? Maybe leveraging the default overload or another overload the . Can we try to register a aten::add.Tensor.default that we don't implement today because of our onnx-fx interpreter can match aten::add.Tensor.default to aten::add.Tensor that already exists?

In this approach, we would have to create that ATen overload ahead in torch side to demonstrate it, which is over complicated the three examples here. Otherwise, we never get the overload from FX.

> Maybe something similar to what is being discussed at https://github.com/pytorch/pytorch/pull/109966 in which aten::split(tensor,dim) already exists but a aten::split(tensor, dim, drop_remainder=False) is added with the same behavior? #109966 could be improved to make this kind of scenario possible, if not already

In that case, user will not be able to overwrite, because schema doesn't match. You have to remove the overload first, and register it with the new schema. Schema is a primary key to the whole matching, kwargs is part of them. 

Overall, I think this tutorial is full of new things, and we should expose overload usages later. Maybe after having some feedback on registry, and then we can go from there.